### ISSUE TITLE
[7.2.0] Emote Menu not working properly

### ISSUE THREAD

**Issue by Px05119** [night/BetterTTV#2901]
Emote menu in 7.2.0 only loads a few emotes in the emote list. Issue exists for both in-window and normal popout chat, legacy popout unaffected (due to older Twitch/BTTV version). Cleared cache, issue still exists.

Side by side of legacy popout (left) and normal popout (right)
![](https://i.gyazo.com/5d230d19bebc7de6b173c2ba23d799f2.png)

**Comment by night** [comment]
unfortunately without proper reproduction steps it's hard to track down the root cause for this issue. i did fix i think 1 case where it was causing the emotes to be out of date on open

**Comment by theSquallo** [comment]
I don't actually know how to reproduce this bug, but only because it's always happens to me on new twitch and ff 57.

**Comment by bspotswood** [comment]
I'm having the same issue as well. I've found that if I disable the BTTV Animated GIF emotes and then refresh the page then I will see everything in the emote menu. Then if I enable the animated emotes and stay there (no refresh) then I will also have them. But I will experience the issue again next time I come back to Twitch if I leave that animated emote setting enabled. I don't have time right now, but might be able to debug it later and see what's actually going on.

**Comment by BluJasmine** [comment]
No emotes are showing up for me at all, even with uninstalling/reinstalling chrome and clearing my cache and cookies.

**Comment by BluJasmine** [comment]
You closed both my comments without any response

**Comment by Boothin** [comment]
I have the same issue with emote list not working, Seems to be an issue with gamewisp emotes. With the gamewisp emote setting enabled, I get 
`WebSocket connection to 'wss://emotes.gamewisp.com/' failed: Error in connection establishment: net::ERR_INSECURE_RESPONSE`
in the console, and emote menu does not load past the BTTV channel emotes. With the gamewisp emote option turned off, the emote menu fully loads everytime.

**Comment by luki1412** [comment]
Yes, this emote menu is still broken. Only loads channel emotes for me. If you go through hosting it still shows emotes from the previous channel.

**Comment by bspotswood** [comment]
I debugged it and found that it's a race condition going on between when UIMenu is initailized and when the "load.chat" watch is triggered. This issue happens when "load.chat" occurs BEFORE the UIMenu is initialized and as the "dom" property of itself set (from it's init() method). 

The way the UIMenu object gets initialized is through a funky setInterval which is commented with "lol this is stupid.. i know". This is found inside the "twitch-chat-emotes" module's main "script.js" file. It uses a 1s interval to try and locate "$('.chat-input button[data-a-target="chat-settings"]')[0]". Once it finds that, and it's a new button, then it calls into a function which calls into the UIMenu object's init() method.

It appears that the "stupid" code above is being used to handle navigation changes on Twitch, detecting a new chat button coming into existence after navigating, and then initializing.

The detection for "load.chat" inside of the watcher module is using a much faster interval of 25ms. I'm not sure why some people such as myself are experiencing the issue while others apparently are not. Perhaps just faster/slower computers or more chrome extensions or something is leading to the varying experiences.

It really seems like the issue pretty much lies in weakness of the twitch-chat-emotes module. I think it needs to have this "stupid" code reworked and it might be a good idea to expose a better mechanism from that module for raising an event when it is ready for 3rd party plugin activity, and/or adding safeguards to ensure that the UIMenu dom is available to append to, even if it isn't yet attached into the actual page dom (no parent yet). Until it is addressed in that module, I don't know that it's worth trying to solution any fix or workaround in BTTV.

**Comment by bspotswood** [comment]
@night - I created a fix to your fork of the module and submitted a pull request for it. I tested it a bit, going between channels, reloading the page, hiding/showing chat, popping it out, and the legacy popout. Seems everything is working with this fix.

https://github.com/night/Userscript--Twitch-Chat-Emotes/pull/4

**Comment by TheGloriousArgonaut** [comment]
@bspotswood I just installed that fix and still notice completely blank emote menus. I have greasemonkey and am using firefox.

**Comment by bspotswood** [comment]
@TheGloriousArgonaut I'm not sure how exactly you're installing that fix in greasemonkey and running it. Although it's difficult to test in FireFox in the same way that we test BetterTTV in chrome (FF54+ really doesn't like this self signed SSL setup), I can still use GM to inject it to load via an http call and everything basically works (although the BTTV icons won't load because they fail at loading it via HTTPS). The fix appears to work fine for me. I can easily demonstrate it not working live in FF and then inject this fixed version and the full menu always shows up. In additional testing, I did realize that favorites weren't working, so I've committed an fixes for that now.

Here's the GM script I used to inject it and test in FireFox. Note that this is also as a development setup for BTTV (using "gulp watch" after having replaced the "script.min.js" under "twitch-chat-emotes" and ensuring that the real bttv isn't loading by changing the "hosts" file as described on the BTTV readme).

`
// ==UserScript==
// @name     BetterTTV
// @version  1
// @grant    none
// @run-at document-idle
// @match https://www.twitch.tv/*
// ==/UserScript==

console.log(window.location);
(function(){
  var s = document.createElement('script');
  s.type = "text/javascript"
  s.src = "http://127.0.0.1/betterttv.js"
  document.head.appendChild(s);
})();
`
